### ISSUE TITLE
Ability to toggle number of levels to hide with toggle-quoted

### ISSUE THREAD

**Issue by darrenkenny** [neomutt/neomutt#1802]
## What would you like to be able to do?

Toggle quoted is very useful for long e-mails, but without the context of the lines above
when people response in-line, it becomes pretty useless.

## What needs to change in NeoMutt?

It would be great if it was possible to specify the level of quotes to be hidden, or a number
of lines of context to be maintained when calling `<toggle-quoted>`.

## Have you tried any alternative methods?

I have tried setting the quote_regex to skip the first level, so something like:

```
set quote_regex="^[|>:}#][ \t]?([ \t]*[|>:}#])+"
```

And that gets me some of the way there, but then I lose the colouring of the quote levels,
and anything else that uses the quote level.

**Comment by flatcap** [comment]
Yes, that sounds pretty useful.

It could be a little tricky to implement -- the pager code is already quite complicated.

**Comment by darrenkenny** [comment]
I just took a quick scan through pager.c, and it seems that, with the right code to get a config value, etc. to obtain a value for `toggle_quote_from_level` that it might be as simple
as the change:

```diff
diff --git a/pager.c b/pager.c
index 76535a9232d3..0d59e7b27d3e 100644
--- a/pager.c
+++ b/pager.c
@@ -1711,7 +1711,8 @@ static int display_line(FILE *fp, LOFF_T *last_pos, struct Line **line_info,
     }

     /* this also prevents searching through the hidden lines */
-    if ((flags & MUTT_HIDE) && ((*line_info)[n].type == MT_COLOR_QUOTED))
+    if ((flags & MUTT_HIDE) && ((*line_info)[n].type == MT_COLOR_QUOTED) &&
+       ((*line_info)[n].quoted->index >= toggle_quote_from_level )
       flags = 0; /* MUTT_NOSHOW */
   }

```

Does that look right? If so, if I can get the cycles, I'll take a stab at implementing it.

Any hints on how to create such a config variable, etc?

**Comment by flatcap** [comment]
Oh!  I revise my estimate to: "not as tricky as I thought" :-)

> Does that look right?

Yes, that looks reasonable.

> Any hints on how to create such a config variable?

That's easy enough...
- Define a new Config Item in `init.h`, of type `DT_NUMBER`
- Add a new global variable to `globals.h`
  - type `short`
  - name `C_` + config name converted to MixedCase

> I'll take a stab at implementing it

Thank you very much.
Any questions, just ask.

**Comment by flatcap** [comment]
Ping!

**Comment by darrenkenny** [comment]
FYI, I'm trying this on MacOS, but not having much luck since neomutt keeps crashing on startup for me when I build from the master branch.

I've tried going back a good number of commits, but it's still crashing on a double-free. The trace I'm seeing is:

```
(lldb) run
Process 95137 launched: '/Users/darrenk/Workspaces/External/neomutt/neomutt' (x86_64)
neomutt(95137,0x1001fa5c0) malloc: *** error for object 0x100815a00: pointer being freed was not allocated
neomutt(95137,0x1001fa5c0) malloc: *** set a breakpoint in malloc_error_break to debug
Process 95137 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
    frame #0: 0x00007fff79e9f2c6 libsystem_kernel.dylib`__pthread_kill + 10
libsystem_kernel.dylib`__pthread_kill:
->  0x7fff79e9f2c6 <+10>: jae    0x7fff79e9f2d0            ; <+20>
    0x7fff79e9f2c8 <+12>: movq   %rax, %rdi
    0x7fff79e9f2cb <+15>: jmp    0x7fff79e99457            ; cerror_nocancel
    0x7fff79e9f2d0 <+20>: retq
Target 0: (neomutt) stopped.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
  * frame #0: 0x00007fff79e9f2c6 libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff79f5abf1 libsystem_pthread.dylib`pthread_kill + 284
    frame #2: 0x00007fff79e096a6 libsystem_c.dylib`abort + 127
    frame #3: 0x00007fff79f18077 libsystem_malloc.dylib`malloc_vreport + 545
    frame #4: 0x00007fff79f17e38 libsystem_malloc.dylib`malloc_report + 151
    frame #5: 0x00000001000d1c5b neomutt`mutt_mem_free(ptr=0x000000010100a698) at memory.c:75:5 [opt]
    frame #6: 0x00000001000b48e5 neomutt`ssl_socket_close(conn=0x000000010100a000) at ssl.c:1432:5 [opt]
    frame #7: 0x00000001000b46ee neomutt`ssl_socket_close_and_restore(conn=0x000000010100a000) at ssl.c:650:12 [opt]
    frame #8: 0x00000001000b2f50 neomutt`mutt_socket_close(conn=0x000000010100a000) at socket.c:105:10 [opt]
    frame #9: 0x00000001000a8c93 neomutt`imap_open_connection [inlined] imap_close_connection(adata=<unavailable>) at imap.c:944:5 [opt]
    frame #10: 0x00000001000a8c81 neomutt`imap_open_connection(adata=0x0000000100706d20) at imap.c:929 [opt]
    frame #11: 0x00000001000abb54 neomutt`imap_login(adata=0x0000000100706d20) at imap.c:1900:5 [opt]
    frame #12: 0x00000001000abad6 neomutt`imap_ac_add(a=<unavailable>, m=<unavailable>) at imap.c:1861:9 [opt]
    frame #13: 0x000000010004f01a neomutt`mx_ac_add(a=<unavailable>, m=<unavailable>) at mx.c:1566:7 [opt]
    frame #14: 0x0000000100033aca neomutt`parse_mailboxes(buf=0x00007ffeefbfd2e8, s=0x0000000100800210, data=0, err=<unavailable>) at init.c:1368:9 [opt]
    frame #15: 0x0000000100038423 neomutt`mutt_parse_rc_line(line=<unavailable>, token=<unavailable>, err=<unavailable>) at init.c:3245:14 [opt]
    frame #16: 0x0000000100037ff5 neomutt`source_rc(rcfile_path=<unavailable>, err=0x00007ffeefbfd7a0) at init.c:818:15 [opt]
    frame #17: 0x0000000100037790 neomutt`mutt_init(skip_sys_rc=false, commands=0x00007ffeefbfe5b8) at init.c:3151:11 [opt]
    frame #18: 0x000000010003d7e3 neomutt`main(argc=<unavailable>, argv=0x00007ffeefbfef30, envp=<unavailable>) at main.c:701:7 [opt]
    frame #19: 0x00007fff79d643d5 libdyld.dylib`start + 1
(lldb)
```

I see a lot of changes w.r.t. renaming and free, so not quite sure where to start here.

Any suggestions...

**Comment by gahr** [comment]
Can you post the following from your config file?

* your `folder` setting
* any `ssl_*` setting
* any `mailboxes` commands

Something goes wrong wrt when parsing `mailboxes` and setting up a connection to your IMAP server.

You can also put a breakpoint at `imap_open_connection` and follow from there. Or enable logging with `-d4` and see whether anything interesting gets dumped to your log file.

**Comment by darrenkenny** [comment]
Hmm, I ran with -d4, and noticed that my old ~/.mutt/notmuch.rc file was being sourced, and other *.rc files in ~/.mutt, when they were commented out in my ~/.muttrc file.

So I compared this with my running configuration, from MacPorts, and it seems that doesn't happen there.

So when I moved my notmuch.rc file to one side, things started working again! I've not been using that configuration for a while now, but it never presented an issue before.

Since when does neomutt source any .rc file directly in ~/.mutt/ ? That is a little confusing, since these are then being sourced out of order from how I'm calling them in my .muttrc.

**Comment by gahr** [comment]
This is the algorithm to locate a config file: https://github.com/neomutt/neomutt/blob/master/init.c#L315-L330

**Comment by darrenkenny** [comment]
OK, very confused here - I definitely saw notmuch in the log file, but when I make sure I'm in the correct directory, and running the correct built binary, it never gets past the first line of the log before the crash.

Seems lldb ./neomutt will run neomutt from the PATH if not in the current directory, unexpectedly.

Sorry for the confusion.

Anyway, still not getting a running binary, and still crashing, I've looked at my configuration, and now that I think about it, I'm not even using SSL for my IMAP connection, it's to a local IMAP server on my local machine using `folder='imap://localhost:143/'`.

So, I recompiled with no optimisation - the config keeps adding a -O2, and was able to debug a little more, and found that the variable `C_SslForceTls` is defaulting to true, causing `imap_open_connection` to bail out, and end up where we are, on closing an ssl connection that was never opened with ssl in the first place.

If I then set the configuration variable `ssl_force_tls` to `no` things start working form me, previously this was unset.

This seems to have been changed by:

    b7ddd2f0d315e3 ("Turn on $ssl_force_tls by default")

but seems like it is not being well handled in the edge case suggested in the
commit log, where I'm using a "a localhost IMAP server with no cert" :)

Will try to fix my original issue when I get a chance now that things are
working again for me, and not crashing.
